core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -313,24 +323,66 @@ private static SEXP handleS4object(@Current Context context, SEXP source, PairLi
     // arguments. In case signature is shorted than the number of arguments we don't need to evaluate the extra
     // arguments.
     
-    int signatureLength = computeSignatureLength(genericMethodEnvironment, groupMethodEnvironment);
+    int[] signatureLength = computeSignatureLength(genericMethodEnvironment, groupMethodEnvironment);
     
-    if(signatureLength == 0) {
+    int maxSignatureLength = 0;
+    for(int i = 0; i < signatureLength.length; i++) {
+      maxSignatureLength = signatureLength[i] < maxSignatureLength ? maxSignatureLength : signatureLength[i];
+    }
+    if(maxSignatureLength == 0) {
       return null;
     }
     
-    List<MethodRanking> possibleSignatures = generatePossibleSignatures(context, rho, genericMethodEnvironment, groupMethodEnvironment, args, signatureLength);
-    
-    Collections.sort(possibleSignatures);
-    
-    List<SelectedMethod> selectedMethods = findMatchingMethods(context, genericMethodEnvironment, groupMethodEnvironment, possibleSignatures);
+    List<Promise> promisedArgs = new ArrayList<>();
+    PairList.Builder rePromisedArgs = new PairList.Builder();
+    Iterator<PairList.Node> it = args.nodes().iterator();
+    int argIdx = 0;
+    while(it.hasNext()) {
```

Comment:
one PairList (inputArgs) is used for argument matching and another one (rePromisedArgs) as arguments to return ClosureDispatch. This only worked if I created the inputArgs from Promise's stored in a intermediate list (promisedArgs), otherwise different test cases failed.